org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/TryWithResourcesJavacFilter.java

```diff
@@ -0,0 +1,245 @@
+
+			AbstractInsnNode startOnNonExceptionalPath = start.getPrevious();
+			cursor = startOnNonExceptionalPath;
+			while (!nextIsJavacClose(p, "n")) {
+				startOnNonExceptionalPath = startOnNonExceptionalPath
+						.getPrevious();
+				cursor = startOnNonExceptionalPath;
+				if (cursor == null) {
+					return false;
+				}
+			}
+			startOnNonExceptionalPath = startOnNonExceptionalPath.getNext();
+
+			final AbstractInsnNode m = cursor;
+			next();
+			if (cursor.getOpcode() != Opcodes.GOTO) {
+				cursor = m;
+			}
+
+			output.ignore(startOnNonExceptionalPath, cursor);
+			output.ignore(start, end);
+			return true;
+		}
+
+		/**
+		 * On a first invocation will associate variables with names "r" and
+		 * "primaryExc", on subsequent invocations will use those associations
+		 * for checks.
+		 */
+		private boolean nextIsJavacClose(final JavacPattern p,
+				final String ctx) {
+			switch (p) {
+			case METHOD:
+			case FULL:
+				// "if (r != null)"
+				if (!(nextIsVar(Opcodes.ALOAD, "r")
+						&& nextIs(Opcodes.IFNULL))) {
+					return false;
+				}
+			}
+			switch (p) {
+			case METHOD:
+			case OPTIMAL:
+				if (nextIsVar(Opcodes.ALOAD, "primaryExc")
+						&& nextIsVar(Opcodes.ALOAD, "r")
+						&& nextIs(Opcodes.INVOKESTATIC)) {
+					final MethodInsnNode m = (MethodInsnNode) cursor;
+					return "$closeResource".equals(m.name)
+							&& "(Ljava/lang/Throwable;Ljava/lang/AutoCloseable;)V"
+									.equals(m.desc);
+				}
+				return false;
+			case FULL:
+			case OMITTED_NULL_CHECK:
+				return nextIsVar(Opcodes.ALOAD, "primaryExc")
+						// "if (primaryExc != null)"
+						&& nextIs(Opcodes.IFNULL)
+						// "r.close()"
+						&& nextIsClose() && nextIs(Opcodes.GOTO)
+						// "catch (Throwable t)"
+						&& nextIsVar(Opcodes.ASTORE, ctx + "t")
+						// "primaryExc.addSuppressed(t)"
+						&& nextIsVar(Opcodes.ALOAD, "primaryExc")
+						&& nextIsVar(Opcodes.ALOAD, ctx + "t")
+						&& nextIsAddSuppressed() && nextIs(Opcodes.GOTO)
+						// "r.close()"
+						&& nextIsClose();
+			default:
+				return false;
+			}
+		}
+
+		private boolean nextIsClose() {
+			if (!nextIsVar(Opcodes.ALOAD, "r")) {
+				return false;
+			}
+			next();
+			if (cursor.getOpcode() != Opcodes.INVOKEINTERFACE
+					&& cursor.getOpcode() != Opcodes.INVOKEVIRTUAL) {
+				return false;
+			}
+			final MethodInsnNode m = (MethodInsnNode) cursor;
+			if (!"close".equals(m.name) || !"()V".equals(m.desc)) {
+				return false;
+			}
+			final String actual = m.owner;
+			if (expectedOwner == null) {
+				expectedOwner = actual;
+				return true;
+			} else {
+				return expectedOwner.equals(actual);
+			}
+		}
+
+		private boolean nextIsAddSuppressed() {
+			if (!nextIs(Opcodes.INVOKEVIRTUAL)) {
+				return false;
+			}
+			final MethodInsnNode m = (MethodInsnNode) cursor;
+			return "java/lang/Throwable".equals(m.owner)
+					&& "addSuppressed".equals(m.name);
+		}
+
+		private boolean nextIsVar(final int opcode, final String name) {
```

Comment:
@marchof can extract common part of two filters (`cursor`, `nextIs`, `nextIsVar`) into abstract class to get rid of duplication and reduce their size/complexity. WDYT?